pkg/controller/expand_and_recover.go

```diff
@@ -0,0 +1,232 @@
+	// we must not try expansion here
+	if ctrl.usedPVCs.hasInUseErrors(pvc) && ctrl.usedPVCs.checkForUse(pvc) {
+		// Record an event to indicate that resizer is not expanding the pvc
+		msg := fmt.Sprintf("Unable to expand %s because CSI driver %s only supports offline expansion and volume is currently in-use", util.PVCKey(pvc), ctrl.resizer.Name())
+		ctrl.eventRecorder.Event(pvc, v1.EventTypeWarning, util.VolumeResizeFailed, msg)
+		return pvc, pv, fmt.Errorf(msg), false
+	}
+
+	// Record an event to indicate that external resizer is resizing this volume.
+	ctrl.eventRecorder.Event(pvc, v1.EventTypeNormal, util.VolumeResizing,
+		fmt.Sprintf("External resizer is resizing volume %s", pv.Name))
+
+	// before trying expansion we will remove the PVC from map
+	// that tracks PVCs which can't be expanded when in-use. If
+	// pvc indeed can not be expanded when in-use then it will be added
+	// back when expansion fails with in-use error.
+	ctrl.usedPVCs.removePVCWithInUseError(pvc)
+	pvc, pv, err = ctrl.callResizeOnPlugin(pvc, pv, newSize, pvcStatusSize)
+
+	if err != nil {
+		// Record an event to indicate that resize operation is failed.
+		ctrl.eventRecorder.Eventf(pvc, v1.EventTypeWarning, util.VolumeResizeFailed, err.Error())
+		return pvc, pv, err, true
+	}
+
+	klog.V(4).Infof("Update capacity of PV %q to %s succeeded", pv.Name, newSize.String())
+	return pvc, pv, nil, true
+}
+
+func (ctrl *resizeController) callResizeOnPlugin(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume,
+	newSize, oldSize resource.Quantity) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error) {
+	updatedSize, fsResizeRequired, err := ctrl.resizer.Resize(pv, newSize)
+
+	if err != nil {
+		// if this error was a in-use error then it must be tracked so as we don't retry without
+		// first verifying if volume is in-use
+		if inUseError(err) {
+			ctrl.usedPVCs.addPVCWithInUseError(pvc)
+		}
+		if isFinalError(err) {
+			var markExpansionFailedError error
+			pvc, markExpansionFailedError = ctrl.markControllerExpansionFailed(pvc)
+			if markExpansionFailedError != nil {
+				return pvc, pv, fmt.Errorf("resizing failed in controller with %v but failed to update PVC %s with: %v", err, util.PVCKey(pvc), markExpansionFailedError)
+			}
+		}
+		return pvc, pv, fmt.Errorf("resize volume %q by resizer %q failed: %v", pv.Name, ctrl.name, err)
+	}
+
+	klog.V(4).Infof("Resize volume succeeded for volume %q, start to update PV's capacity", pv.Name)
+
+	pv, err = ctrl.updatePVCapacity(pv, oldSize, updatedSize, fsResizeRequired)
+	if err != nil {
+		return pvc, pv, fmt.Errorf("error updating pv %q by resizer: %v", pv.Name, err)
+	}
+
+	if fsResizeRequired {
+		pvc, err = ctrl.markForPendingNodeExpansion(pvc)
+		return pvc, pv, err
+	}
+	pvc, err = ctrl.markOverallExpansionAsFinished(pvc, updatedSize)
+	return pvc, pv, err
+}
+
+// checks if pv can be expanded
+func (ctrl *resizeController) pvCanBeExpanded(pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim) bool {
```

Comment:
I can do that, but personally I do not see much value in touching older functions. I would like to remove all of those functions once this feature goes beta in 1.24 (which is just few months away). So that is why, I have done side-by-side implementation rather than adding feature-gate checks all over the place, this way I can simply remove those functions when this feature goes beta.

Reply:
I think it will need to be there until GA, the feature still can be disabled by a feature gate.

Reply:
ack - although existing implementation is beta too, but I think we can discuss this later.